esm: add `initialize` hook, integrate with `register`

izaakschroeder opened the pull request:
Follows @giltayar's proposed API:

> `register` can pass any data it wants to the loader, which will be passed to the exported `initialize` function of the loader. Additionally, if the user of `register` wants to communicate with the loader, it can just create a `MessageChannel` and pass the port to the loader as data.

The `register` API is now:

```ts
interface Options {
  parentUrl?: string;
  data?: any;
  transferList?: any[];
}

function register(loader: string, parentUrl?: string): any;
function register(loader: string, options?: Options): any;
```

This API is backwards compatible with the old one (new arguments are optional and at the end) and allows for passing data into the new `initialize` hook. If this hook returns data it is passed back to `register`:

```ts
function initialize(data: any): Promise<any>;
```

**NOTE**: Currently there is no mechanism for a loader to exchange ownership of something back to the caller.

Refs: https://github.com/nodejs/loaders/issues/147

----

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

izaakschroeder commented:
/cc @@targos @giltayar

anonrig commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -127,28 +127,31 @@ class Hooks {
    * Import and register custom/user-defined module loader hook(s).
    * @param {string} urlOrSpecifier
    * @param {string} parentURL
+   * @param {any} data
```
Can you add a more verbose description about this parameter?

anonrig commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -127,28 +127,31 @@ class Hooks {
       kEmptyObject,
     );
-    this.addCustomLoader(urlOrSpecifier, keyedExports);
+    return await this.addCustomLoader(urlOrSpecifier, keyedExports, data);
```
```suggestion
    return this.addCustomLoader(urlOrSpecifier, keyedExports, data);
```

izaakschroeder replied:
Is there a reason it's preferred to avoid `await`ing async calls in async functions if they are the terminal call?

ljharb replied:
it's redundant and adds an artificially forced tick, thus making it slower.

izaakschroeder replied:
TIL. Thanks!

anonrig commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -162,6 +165,9 @@ class Hooks {
       ArrayPrototypePush(this.#chains.load, { __proto__: null, fn: load, url, next });
     }
+    if (initialize) {
+      return await initialize(data);
```
```suggestion
      return initialize(data);
```

anonrig commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -553,15 +559,18 @@ class HooksProxy {
   }
 
-  async makeAsyncRequest(method, ...args) {
+  async makeAsyncRequest(transferList, method, ...args) {
```
Can you add a JSDoc?

anonrig commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -553,15 +559,18 @@ class HooksProxy {
+    debug('post async message to worker', { method, args, transferList });
+    const finalTransferList = transferList ?
+      [asyncCommChannel.port2, ...transferList] :
+      [asyncCommChannel.port2]
```
```suggestion
      [asyncCommChannel.port2];
```

anonrig commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -593,12 +602,12 @@ class HooksProxy {
   }
 
-  makeSyncRequest(method, ...args) {
+  makeSyncRequest(transferList, method, ...args) {
```
Can you add a jsdoc?

ljharb commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -593,12 +602,12 @@ class HooksProxy {
-    debug('post sync message to worker', { method, args });
-    this.#worker.postMessage({ method, args });
+    debug('post sync message to worker', { method, args, transferList });
+    this.#worker.postMessage({ method, args }, transferList);
```
```suggestion
    this.#worker.postMessage({ __proto__: null, method, args }, transferList);
```
(altho this predates this PR)

ljharb commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -553,15 +559,18 @@ class HooksProxy {
+    const finalTransferList = transferList ?
+      [asyncCommChannel.port2, ...transferList] :
+      [asyncCommChannel.port2]
+    this.#worker.postMessage({ method, args, port: asyncCommChannel.port2 }, finalTransferList);
```
```suggestion
    this.#worker.postMessage({ __proto__: null, method, args, port: asyncCommChannel.port2 }, finalTransferList);
```
altho this might not matter if the worker's not in object mode and if serialization ignores non-own properties

GeoffreyBooth commented:
> This API is backwards compatible with the old one (new arguments are optional and at the end)

`register` hasn’t been released yet, so we don’t need to preserve any backward compatibility. That said, the proposed API seems fine. (Do we need `parentURL`? It seems like it’s already optional and set to `process.cwd` if undefined, so maybe we can do without it and users can pass as the first argument a path to the loader relative to CWD or an absolute URL to the loader.)

The docs need updating as part of this PR. That would help in reviews, as we can see the intended usage from the example in the new docs. Since `register` should work in CommonJS too, it would be good to have both CommonJS and ESM versions of the examples.

aduh95 commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -127,28 +127,31 @@ class Hooks {
+   * @param {any} data Arbitrary data to pass to loader's `initialize`
    */
-  addCustomLoader(url, exports) {
+  async addCustomLoader(url, exports, data) {
```
Why `async`?

aduh95 commented on test/fixtures/es-module-loaders/hooks-initialize.mjs:
```diff
@@ -0,0 +1,3 @@
+export async function initialize() {
+  console.log('hooks initialize');
+}
```
```suggestion
}

```

aduh95 commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -162,6 +165,9 @@ class Hooks {
       const next = this.#chains.load[this.#chains.load.length - 1];
       ArrayPrototypePush(this.#chains.load, { __proto__: null, fn: load, url, next });
     }
+    if (initialize) {
+      return initialize(data);
+    }
```
```suggestion
    return initialize?.(data);
```

aduh95 commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -553,15 +559,25 @@ class HooksProxy {
-    debug('post async message to worker', { method, args });
-    this.#worker.postMessage({ method, args, port: asyncCommChannel.port2 }, [asyncCommChannel.port2]);
+    debug('post async message to worker', { method, args, transferList });
+    const finalTransferList = transferList ?
+      [asyncCommChannel.port2, ...transferList] :
+      [asyncCommChannel.port2];
```
```suggestion
    const finalTransferList = [asyncCommChannel.port2];
    if (transferList) {
      ArrayPrototypePushApply(finalTransferList, transferList);
    }
```

aduh95 commented on lib/internal/modules/esm/loader.js:
````diff
@@ -497,18 +499,22 @@ function getHooksProxy() {
  * ```
  */
-function register(specifier, parentURL = 'data:') {
+function register(specifier, parentURL = 'data:', data = undefined, transferList = []) {
````
Let's not create a throw away array and keep the `transferList` as undefined instead.
```suggestion
function register(specifier, parentURL = 'data:', data = undefined, transferList) {
```

izaakschroeder commented:
Thanks for all the feedback! Work is a bit busy, but I will get to this later tonight (Friday) 😄

izaakschroeder commented:
> The docs need updating as part of this PR. That would help in reviews, as we can see the intended usage from the example in the new docs. Since register should work in CommonJS too, it would be good to have both CommonJS and ESM versions of the examples.

@GeoffreyBooth how/where do I update the docs?

GeoffreyBooth commented:
> how/where do I update the docs?

Edit the `module.md` file in the repo: https://github.com/nodejs/node/blob/main/doc/api/module.md#moduleregister

You’ll find loaders documented in `esm.md`: https://github.com/nodejs/node/blob/main/doc/api/esm.md#loaders

izaakschroeder commented:
@GeoffreyBooth while I do up the docs, is the CJS test here sufficient or do you want to see something more?

GeoffreyBooth commented:
> is the CJS test here sufficient or do you want to see something more?

We should add a test that specifically uses `--require`, not just `--eval`, as the former is a different code path that runs at a different point in startup. Basically, I expect some users to run commands like `node --require ts-node/register app.js` and that should work as expected. Bonus points for setting up a `require.extensions` hook that does something and it's coordinated with the `load` hook. That simulates the transpilation use case.

izaakschroeder commented:
@GeoffreyBooth I added two more tests but one of them is failing. I added `{skip: ...}` for now but I will have to do further investigation. Let me know if you have any ideas on the source of this failure 😄

GeoffreyBooth commented:
Please don't add `skip`, we don't want to accidentally land a PR with failing tests. It's okay if CI fails while we're still working.

izaakschroeder commented:
Oh, sorry, removing skip 😅

JakobJingleheimer commented on lib/internal/modules/esm/loader.js:
```diff
@@ -297,15 +297,15 @@ class ModuleLoader {
   }
 
-  register(specifier, parentUrl) {
+  register(specifier, parentUrl, data, transferList) {
```
```suggestion
  /**
   * @see {CustomizedModuleLoader.register}
   */
  register(specifier, parentUrl, data, transferList) {
```

I think this syntax is right

aduh95 replied:
According to https://jsdoc.app/tags-see.html, it should be
```suggestion
  /**
   * @see {@link CustomizedModuleLoader.register}
   */
  register(specifier, parentUrl, data, transferList) {
```

Not sure if it matters though as we are not generating our docs with the JSDoc comments.

JakobJingleheimer replied:
VS Code (and possibly others) consumes them :)

JakobJingleheimer commented on lib/internal/modules/esm/loader.js:
```diff
@@ -497,18 +499,22 @@ function getHooksProxy() {
  * @param {string} [parentURL]
- * @returns {void}
+ * @param {any} [data] Arbitrary data passed to loader's `initialize` hook
+ * @param {any[]} transferList Objects in `data` that are changing ownership
```
If `data` is optional, `transferList must also be optional:

```suggestion
 * @param {any[]} [transferList] Objects in `data` that are changing ownership
```

JakobJingleheimer commented on test/es-module/test-esm-loader-hooks.mjs:
```diff
@@ -553,4 +553,140 @@ describe('Loader hooks', { concurrency: true }, () => {
+      '--eval',
+      `
+        import os from 'node:os';
+        import fs from 'node:fs';
```
Note that FS gets mocked, so it's best to avoid this unless you want the mock.

izaakschroeder replied:
I can probably remove that line entirely tbh. As long as the initialize hook is triggered is all I care about in this test.

JakobJingleheimer reviewed (commented):
This is looking good!

If you'd like help with anything, I have some time this weekend (and then I'm away starting Wednesday for ~2 weeks)

JakobJingleheimer commented:
> Do we need `parentURL`?

@GeoffreyBooth yes: this addresses a specific need; I can't remember what it is, but I'm sure it was discussed. I can try to find it if you'd like.

aduh95 commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -553,15 +558,26 @@ class HooksProxy {
-    this.#worker.postMessage({ method, args, port: asyncCommChannel.port2 }, [asyncCommChannel.port2]);
+    debug('post async message to worker', { method, args, transferList });
+    const finalTransferList = [asyncCommChannel.port2];
+    if (transferList) {
+      ArrayPrototypePushApply(finalTransferList, transferList);
+    }
```
I'm pretty sure this wouldn't work, it would throw something like `Value at index 1 does not have a transferable type.`

JakobJingleheimer replied:
🤯 why?

aduh95 replied:
Because arrays are not transferable (see https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Transferable_objects#supported_objects). You can convince yourself by running e.g. `{const a=new ArrayBuffer(9);structuredClone({a}, {transfer:[[a]]})}`.

JakobJingleheimer replied:
Ohhh! Whoops this is `push.apply`, not `push`. Nvmd, they're not equivalent.

GeoffreyBooth commented:
> @GeoffreyBooth yes: this addresses a specific need; I can’t remember what it is, but I’m sure it was discussed. I can try to find it if you’d like.

Yes, if you can find the reason, I’d appreciate it. I assume it’s because we can’t figure out how to search for `specifier` relative to the importing file. Assuming that’s the reason, I still think I’d like to remove `parentUrl` from the signature. It defaults to `process.cwd`, so I think we could eliminate it and anyone who _would_ have been passing it (because `specifier` is relative to some base other than CWD) can instead call `register(new URL(specifier, parentURL).href, data, transferList)`.

And if `specifier` cannot be found, the error message should say something about Node searching for it _relative to the current working directory,_ rather than relative to the importing file. I think we should probably add this extra error text regardless of whether or not `parentUrl` stays or goes, as I think this “relative to CWD” behavior is atypical for Node and potentially surprising.

How do others feel about this?

targos commented:
I would prefer the signature to be `loader, options?`.

aduh95 commented:
> can instead call `register(new URL(specifier, parentURL).href, data, transferList)`.

You would not get the expected result if you try to register a `node_module` package.

GeoffreyBooth commented:
> You would not get the expected result if you try to register a `node_module` package.

For a package you would just pass the specifier, e.g. `register('ts-node')`. My example was only for custom loaders, like `register(new URL('./my-custom-loader.js', import.meta.url).href)`.

> the only case I can think of where resolving relative to the CWD makes sense is to parse a relative path passed as CLI argument

Yes, except that’s how `register` currently works as far as I know. I assume we all agree that it would be nice if it worked relative to the current file, like how `import()` works, if anyone can figure out how to make that happen.

aduh95 commented:
> For a package you would just pass the specifier, e.g. `register('ts-node')`

Yes, but if you resolve from the CWD, you'll get weird results, that's almost certainly not what the users expect.

> I assume we all agree that it would be nice if it worked relative to the current file, like how `import()` works, if anyone can figure out how to make that happen.

Like you said, we probably all agree, and it seems the only way of making that happen is to pass `import.meta.url` to the `register` method (the `parentURL` argument).

JakobJingleheimer commented:
@GeoffreyBooth yes! that's the reason.

> it seems the only way of making that happen is to pass `import.meta.url` to the `register` method (the `parentURL` argument).

**+1**

This seems a very simple solution for basically no effort, is very easy for users to understand, and doesn't get in the way when not used 🙂

GeoffreyBooth commented:
> Like you said, we probably all agree, and it seems the only way of making that happen is to pass `import.meta.url` to the `register` method (the `parentURL` argument).

What I’m saying is that we shouldn’t require users to do this. `register('./my-loader.js')` should behave the same as `register('./my-loader.js', import.meta.url)`. Requiring `import.meta.url` to be explicitly passed in, rather than implied as a default, feels like a failure on our part.

It’s not something that needs to be fixed in this PR if it’s hard, but I feel like we should try to fix this before we stabilize `register`. If we change the signature to `register(specifier, options?)` as @targos suggests then it would be less disruptive if we find a way to make `options.parentURL` optional.

> if you resolve from the CWD, you’ll get weird results, that’s almost certainly not what the users expect.

I assume that most apps run like `node entry.js` where `entry.js` is at the root of the app, a sibling to `node_modules`. In which case, resolving from the CWD should find a package in `node_modules` as I’d expect. What weird results are you thinking of?

Now maybe it’s not overwhelmingly common that CWD is the same folder that’s the parent of `node_modules`; I don’t know what the edge cases are (apps run via bundlers like Vite? Frameworks like Next?). Hence all the more reason to try to figure out a way to default `parentURL` to `import.meta.url` rather than CWD.

aduh95 commented:
> Hence all the more reason to try to figure out a way to default `parentURL` to `import.meta.url` rather than CWD.

It seems to me you are the only talking about using CWD as the default value, but you also seem to agree that it wouldn't be a good value, there seems to be some kind of misunderstanding here. Currently the default value is `data:`, which was picked so the user gets a nice error if they try to pass a relative specifier without passing an explicit `parentURL`.
I think it's fair to say that parsing error stack trace is not a viable way for getting the `parentURL`, so asking the user to pass it themself is the next best thing.

> I assume that most apps run like `node entry.js` where `entry.js` is at the root of the app, a sibling to `node_modules`. In which case, resolving from the CWD should find a package in `node_modules` as I’d expect. What weird results are you thinking of?

If the `register` call happens from within the `node_modules` directory, it could resolve to a completely different version of the target library. Also, resolving from the CWD is not a thing anywhere in CJS nor ESM except for `--require` and `--import` CLI flags.

Anyway, the consensus seems clearly to be to not do that in that PR, so +1 to that.


> I would prefer the signature to be `loader, options?`.

Should we do `register(specifier[, parentURL][, options])`?

targos commented:
> > I would prefer the signature to be `loader, options?`.

> Should we do `register(specifier[, parentURL][, options])`?

I wouldn't mind too much, but I tend to prefer named parameters, so everything in options.

GeoffreyBooth commented:
Besides the open comments, please don’t forget to add a new section to the docs explaining `initialize`.

izaakschroeder commented:
I think most stuff is cleaned up @GeoffreyBooth @JakobJingleheimer; I'm pushing up the last commit where all tests pass locally but have to sleep before I can see it through. After some amount of thinking, I think I'm more inclined to agree with @targos in that having everything in options:

a) keeps things explicit – you know what it is you're passing in and it's not some implicit no-name argument,
b) avoids useless/confusing junk – consumers who want `register('supermock', {data: ...})` would probably be happier to not have to stuff in a seemingly random `import.meta.url` or some such.
c) if we ever come up with some magic for obviating `import.meta.url` then we can quietly provide a default without interrupting the otherwise normal calling flow.

So this is what I've done. The commit is revertible if you feel strongly either way. We could also `typeof` check to see if `options` is a string for backwards compatibility but you noted that isn't the largest concern here.

I've attempted a start on the documentation… I'm not a particularly prolific writer and could probably use some guidance there… I'm not sure what examples are appropriate and I'm probably not good at matching the tone of the existing documentation.

Also I don't know how to do deprecations… I kind of just winged it based on what I saw in the repo 😄 

Hope you all had a great weekend!

izaakschroeder commented on doc/api/esm.md:
```diff
@@ -938,19 +944,82 @@ export async function load(url, context, nextLoad) {
+> signature may change. Do not rely on the API described below.
+
+> In a previous version of this API, this was split across 3 separate, now
+> deprecated, hooks (`getFormat`, `getSource`, and `transformSource`).
```
Er derp. I need to remove this block 😓

izaakschroeder commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -127,31 +129,44 @@ class Hooks {
     if (globalPreload) {
+      if (globalPreloadWarned === false) {
+        globalPreloadWarned = true;
+        process.emitWarning(
```
How… do I test this? Should I?

aduh95 commented on doc/api/deprecations.md:
```diff
@@ -3380,6 +3380,21 @@ In a future version of Node.js, [`assert.CallTracker`][],
 will be removed.
 Consider using alternatives such as the [`mock`][] helper function.
 
+### DEP0174: the `globalPreload` hook
+
+<!-- YAML
+changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/48842
+    description: Superseded be `initialize`.
+-->
+
+Type: Runtime
+
+In a future version of Node.js, the `globalPreload` hook,
+will be removed. Consider using alternatives such as the
+`initialize` hook.
+
```
I'm -1 on deprecating an experimental API. A deprecated API is expected to follow the semver contract, an experimental API is not.
```suggestion
```

aduh95 commented on doc/api/esm.md:
```diff
@@ -685,6 +685,12 @@ of Node.js applications.
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/48842
+    description: Added `initialize` hook and deprecated `getGlobalPreload`.
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/48559
+    description: Allow loaders to use `import()` and unflag `Module.register`.
```
Can you please move that to its own PR? Otherwise the `REPLACEME` will likely be incorrect depending on when we backport each PR.

aduh95 commented on doc/api/esm.md:
```diff
@@ -938,19 +944,82 @@ export async function load(url, context, nextLoad) {
 #### `globalPreload()`
 
 <!-- YAML
+deprecated: REPLACEME
```
```suggestion
```

aduh95 commented on doc/api/esm.md:
```diff
@@ -938,19 +944,82 @@ export async function load(url, context, nextLoad) {
 
-> The loaders API is being redesigned. This hook may disappear or its
-> signature may change. Do not rely on the API described below.
+> Deprecated: Use `initialize` instead.
```
Can you add a link please?
```suggestion
> This hook will be removed in a future version. Use [`initialize`][] instead.
```

aduh95 commented on doc/api/esm.md:
```diff
@@ -938,19 +944,82 @@ export async function load(url, context, nextLoad) {
 <!-- YAML
+deprecated: REPLACEME
 changes:
+  - version:
+    - REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/48842
+    description: Deprecated in favor of `initialize` hook.
```
```suggestion
```

aduh95 commented on doc/api/module.md:
```diff
@@ -93,7 +93,7 @@ loaders can be registered programmatically using the
 import { register } from 'node:module';
 
-register('http-to-https', import.meta.url);
+register('http-to-https', { parentUrl: import.meta.url });
```
If you pass an absolute URL (that can be file:, data:, or https:), the `parentURL` is not relevant. For node_modules packages resolution and relative URLs, you need one.

izaakschroeder replied:
Well it accepts both now, so everyone can be happy 🎉

aduh95 replied:
Can you please revert this change to minimize the diff?

izaakschroeder commented:
Next round of updates in the bag @aduh95 @GeoffreyBooth @JakobJingleheimer 🎉

aduh95 commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -127,31 +131,42 @@ class Hooks {
+      initialize,
     } = pluckHooks(exports);
 
     if (globalPreload) {
```
We should ensure that the warning won’t be printed if the loader contains both `initialize` and `globalPreload`. Actually we should completely ignore `globalPreload` if `initialize` is provided. 
```suggestion
    if (globalPreload && initialize == null) {
```

izaakschroeder replied:
So the behaviour if they provide both is to ignore `globalPreload` entirely? Should we not at least warn that they're providing both and only one will be used?

aduh95 replied:
The assumption is that if they know about `initialize`, they know it's the replacement for `globalPreload` and they expect only one will be used. We have to do that otherwise there won't be a way to write a loader that works in both versions.

aduh95 commented on lib/internal/process/pre_execution.js:
```diff
@@ -154,9 +154,16 @@ function setupUserModules(isLoaderWorker = false) {
-  initializeFrozenIntrinsics();
+  if (isLoaderWorker) {
+    // Loader workers are responsible for doing this themselves.
+    globalThis.loadPreloadModules = loadPreloadModules;
```
Added globals must be documented in `globals.md`. But we should definitely NOT do that, adding globals would a breaking change.

aduh95 commented on lib/internal/modules/esm/loader.js:
````diff
@@ -513,19 +519,43 @@ function getHooksProxy() {
  * ```
  */
-function register(specifier, parentURL = 'data:') {
+function register(specifier, parentUrl, options) {
````
```suggestion
function register(specifier, parentURL = 'data:', options) {
```

aduh95 commented on test/es-module/test-esm-loader-programmatically.mjs:
```diff
@@ -27,8 +27,8 @@ describe('ESM: programmatically register loaders', { concurrency: true }, () =>
       "import { register } from 'node:module';" +
-      `register(${JSON.stringify('./loader-resolve-passthru.mjs')}, ${JSON.stringify(parentURL)});` +
-      `register(${JSON.stringify('./loader-load-passthru.mjs')}, ${JSON.stringify(parentURL)});` +
+      `register(${JSON.stringify('./loader-resolve-passthru.mjs')}, { parentUrl:  ${JSON.stringify(parentURL)} });` +
+      `register(${JSON.stringify('./loader-load-passthru.mjs')}, { parentUrl: ${JSON.stringify(parentURL)} });` +
```
```suggestion
      `register(${JSON.stringify('./loader-resolve-passthru.mjs')}, ${JSON.stringify({ parentURL })});` +
      `register(${JSON.stringify('./loader-load-passthru.mjs')}, ${JSON.stringify({ parentURL })} });` +
```

GeoffreyBooth commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -593,12 +624,19 @@ class HooksProxy {
+   * @param  {any[]} args Arguments to pass to `method`
+   * @returns {any}
+   */
+  makeSyncRequest(transferList, method, ...args) {
```
This should parallel whatever signature we decide on for `makeAsyncRequest`.

GeoffreyBooth commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -710,6 +748,7 @@ function pluckHooks({
   globalPreload,
   resolve,
   load,
+  initialize,
```
Let’s try to always keep in this order if you don’t mind.

```suggestion
  initialize,
  resolve,
  load,
```

GeoffreyBooth commented on lib/internal/modules/esm/loader.js:
```diff
@@ -513,19 +519,43 @@ function getHooksProxy() {
  * @param {string} specifier
- * @param {string} [parentURL]
- * @returns {void}
+ * @param {string} parentUrl Same as `options.parentUrl`, just inline
+ * @param {object} [options] Options to pass to `register`
+ * @param {string} [options.parentUrl] Parent URL to use when resolving `specifier`
+ * @param {any} [options.data] Arbitrary data passed to loader's `initialize` hook
+ * @param {any[]} [options.transferList] Objects in `data` that are changing ownership
+ * @returns {any}
```
```suggestion
 * @param {string} [parentUrl] Base to use when resolving `specifier`; optional if `specifier` is absolute. Same as `options.parentUrl`, just inline
 * @param {string} [options.parentUrl] Base to use when resolving `specifier`
 * @param {any} [options.data] Arbitrary data passed to loader's `initialize` hook
 * @param {any[]} [options.transferList] Objects in `data` that are changing ownership
 * @returns {any} Returns data returned by `initialize` hook, if one is present
```

anonrig replied:
Adding `undefined` to returns might be helpful

GeoffreyBooth commented on doc/api/esm.md:
```diff
@@ -938,6 +941,62 @@ export async function load(url, context, nextLoad) {
+in the loader's thread when the loader is initialized. This hook can send
+and receive data from a `Module.register` invocation, including ports and
+other transferrable objects.
+
```
What happens if multiple loaders are registered and they each have `initialize` hooks? Presumably they aren’t chained; they just run in sequence, like `globalPreload`? We should probably document this, and include a test that multiple loaders’ `initialize` hooks behave as expected.

GeoffreyBooth reviewed (commented):
This is really close. Great work!

aduh95 commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -127,31 +131,42 @@ class Hooks {
     } = pluckHooks(exports);
 
-    if (globalPreload) {
+    if (globalPreload && !initialize) {
```
No, we should not error, otherwise it would be impossible to write a loader that works in both versions of Node.js. See discussion in https://github.com/nodejs/node/pull/48842#discussion_r1280129328

GeoffreyBooth replied:
I was thinking that they would just branch on detecting the Node version, but sure, ignoring `globalPreload` if `initialize` is exported seems gentler.

aduh95 commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -553,15 +569,30 @@ class HooksProxy {
 
-  async makeAsyncRequest(method, ...args) {
+  /**
+   * Invoke a remote method asynchronously.
+   * @param {any[]|undefined} transferList Objects in `args` to be transferred
+   * @param {string} method Method to invoke
+   * @param  {any[]} args Arguments to pass to `method`
+   * @returns {Promise<any>}
+   */
+  async makeAsyncRequest(transferList, method, ...args) {
```
Using variable arguments is probably more efficient than using arrays, because it's an internal method it would make sense to me to favor performance. It's easy to change after the fact anyway.

GeoffreyBooth replied:
Okay so then to make `method` first would it need to be `makeAsyncRequest(method, transferList, ...args)`, and `transferList` needs to be specified (even if just as `undefined`)? Or `makeAsyncRequest(method, ...args)` and we pull `transferList` out of the first element of `args`?

izaakschroeder replied:
It's a bit ugly looking I agree… but this was the only sensible way to preserve the `"command", arg1, arg2` style syntax that I saw.

anonrig commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -127,31 +131,42 @@ class Hooks {
+    if (globalPreload && !initialize) {
+      if (globalPreloadWarned === false) {
+        globalPreloadWarned = true;
+        emitExperimentalWarning(
```
I believe the correct way of deprecation is through first documentation-only deprecate it and later runtime deprecate it. Is this because of experimental status?

GeoffreyBooth replied:
Yes, deprecation isn’t something that happens for experimental APIs. By definition they’re still changing.

anonrig commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -127,31 +131,42 @@ class Hooks {
    * @param {Record<string, unknown>} exports
+   * @param {any} [data] Arbitrary data to be passed from the custom loader (user-land)
+   * to the worker.
    */
```
Can you add `@return`?

anonrig commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -593,12 +624,19 @@ class HooksProxy {
-  makeSyncRequest(method, ...args) {
+  /**
+   * Invoke a remote method synchronously.
+   * @param {any[]|undefined} transferList Objects in `args` to be transferred
```
```suggestion
   * @param {any[]} [transferList] Objects in `args` to be transferred
```

anonrig commented on lib/internal/modules/esm/loader.js:
```diff
@@ -513,19 +519,43 @@ function getHooksProxy() {
+function register(specifier, parentUrl, options) {
   const moduleLoader = require('internal/process/esm_loader').esmLoader;
-  moduleLoader.register(`${specifier}`, parentURL);
+  if (typeof parentUrl === 'object') {
```
```suggestion
  if (parentUrl != null && typeof parentUrl === 'object') {
```

izaakschroeder commented on doc/api/esm.md:
```diff
@@ -685,6 +685,9 @@ of Node.js applications.
 changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/48842
+    description: Added `initialize` hook and deprecated `getGlobalPreload`.
```
I did this before and @aduh95 doesn't want this so I reverted my changes :\

aduh95 replied:
Let's remove the term deprecated since it's not what we are technically doing.
```suggestion
    description: Added `initialize` hook to replace `globalPreload`.
```

izaakschroeder commented:
I think this now covers just about everything everyone has asked for @GeoffreyBooth 😄

GeoffreyBooth commented on doc/api/esm.md:
```diff
@@ -742,6 +742,62 @@ different [realm](https://tc39.es/ecma262/#realm). The hooks thread may be
+The `initialize` hook provides a way to define a custom function that runs
+in the loader's thread when the loader is initialized. This hook can send
+and receive data from a `register` invocation, including ports and other
+transferrable objects.
```
```suggestion
transferrable objects. The return value of `initialize` must be something that
can be posted as a message between threads, like the input to
[port.postMessage][].
```

With footnote to https://nodejs.org/api/worker_threads.html#portpostmessagevalue-transferlist. (I’m assuming that this addition is correct?)

izaakschroeder replied:
Yes. `data` as a parameter can be ports and other fun things, but right now the return value has no such magic, unless we wish to change the signature to `return {data, transferList}` or some such.

GeoffreyBooth commented on lib/internal/modules/esm/hooks.js:
```diff
@@ -553,15 +570,30 @@ class HooksProxy {
+   * @param  {any[]} args Arguments to pass to `method`
+   * @returns {Promise<any>}
+   */
+  async makeAsyncRequest(transferList, method, ...args) {
```
This still needs refactoring so that `method` is first again. Likewise for `makeSyncRequest` and other places where `transferList` was added.

izaakschroeder replied:
Ah ok sorry I didn't realize that had been decided on. I can do that though.

aduh95 commented on test/es-module/test-esm-loader-programmatically.mjs:
```diff
@@ -22,12 +22,15 @@ const commonEvals = {
       "import { register } from 'node:module';" +
       commonEvals.register(fixtures.fileURL('es-module-loaders', 'loader-resolve-passthru.mjs')) +
       commonEvals.register(fixtures.fileURL('es-module-loaders', 'loader-load-passthru.mjs')) +
+      `register(${JSON.stringify('./loader-resolve-passthru.mjs')}, { parentURL:  ${JSON.stringify(parentURL)} });` +
+      `register(${JSON.stringify('./loader-load-passthru.mjs')}, { parentURL: ${JSON.stringify(parentURL)} });` +
```
nit
```suggestion
      `register(${JSON.stringify('./loader-resolve-passthru.mjs')}, ${JSON.stringify({ parentURL })});` +
      `register(${JSON.stringify('./loader-load-passthru.mjs')}, ${JSON.stringify({ parentURL })});` +
```

aduh95 commented on doc/api/esm.md:
```diff
@@ -739,6 +742,62 @@ different [realm](https://tc39.es/ecma262/#realm). The hooks thread may be
+> The loaders API is being redesigned. This hook may disappear or its
+> signature may change. Do not rely on the API described below.
+
+* `data` {any} The data from `register(loader, import.meta.url, { data })`
+* Returns: {any} The data to be returned to the caller of `register`
```
```suggestion
* `data` {any} The data from `register(loader, import.meta.url, { data })`.
* Returns: {any} The data to be returned to the caller of `register`.
```

aduh95 commented on doc/api/esm.md:
```diff
@@ -739,6 +742,62 @@ different [realm](https://tc39.es/ecma262/#realm). The hooks thread may be
+const { port1, port2 } = new MessageChannel();
+
+port1.on('message', (msg) => {
+  assert(msg === 'increment: 2');
```
```suggestion
  assert.strictEqual(msg, 'increment: 2');
```

aduh95 commented on doc/api/esm.md:
```diff
@@ -739,6 +742,62 @@ different [realm](https://tc39.es/ecma262/#realm). The hooks thread may be
+  transferList: [port2],
+});
+
+assert(result === 'ok');
```
```suggestion
assert.strictEqual(result, 'ok');
```

aduh95 commented on doc/api/esm.md:
```diff
@@ -949,8 +1008,7 @@ changes:
 
-> The loaders API is being redesigned. This hook may disappear or its
-> signature may change. Do not rely on the API described below.
+> This hook will be removed in a future version. Use [`initialize`][] instead.
```
```suggestion
> This hook will be removed in a future version. Use [`initialize`][] instead.
> When a loader has an `initilize` export, `globalPreload` will be ignored.
```

aduh95 commented on lib/internal/modules/esm/loader.js:
````diff
@@ -513,19 +519,44 @@ function getHooksProxy() {
  * ```
  */
-function register(specifier, parentURL = 'data:') {
+function register(specifier, parentURL, options) {
````
We need to mark the parameter as optional, otherwise `register.length` will report the wrong number of mandatory arguments.
```suggestion
function register(specifier, parentURL = 'data:', options) {
```
or
```suggestion
function register(specifier, parentURL = undefined, options) {
```

aduh95 commented on lib/internal/modules/esm/initialize_import_meta.js:
```diff
@@ -7,14 +7,14 @@ const experimentalImportMetaResolve = getOptionValue('--experimental-import-meta
  * @param {string} defaultParentUrl The default base to use for resolution
  * @param {typeof import('./loader.js').ModuleLoader} loader Reference to the current module loader
- * @returns {(specifier: string, parentUrl?: string) => string} Function to assign to import.meta.resolve
+ * @returns {(specifier: string, parentURL?: string) => string} Function to assign to import.meta.resolve
```
Let's revert that and take it to its own PR since it's not related
```suggestion
 * @returns {(specifier: string, parentUrl?: string) => string} Function to assign to import.meta.resolve
```

aduh95 commented on lib/internal/modules/esm/initialize_import_meta.js:
```diff
@@ -7,14 +7,14 @@ const experimentalImportMetaResolve = getOptionValue('--experimental-import-meta
  */
 function createImportMetaResolve(defaultParentUrl, loader) {
-  return function resolve(specifier, parentUrl = defaultParentUrl) {
+  return function resolve(specifier, parentURL = defaultParentUrl) {
```
Let's revert that and take it to its own PR since it's not related
```suggestion
  return function resolve(specifier, parentUrl = defaultParentUrl) {
```

aduh95 commented on lib/internal/modules/esm/initialize_import_meta.js:
```diff
@@ -7,14 +7,14 @@ const experimentalImportMetaResolve = getOptionValue('--experimental-import-meta
 
     try {
-      ({ url } = loader.resolveSync(specifier, parentUrl));
+      ({ url } = loader.resolveSync(specifier, parentURL));
```
Let's revert that and take it to its own PR since it's not related
```suggestion
      ({ url } = loader.resolveSync(specifier, parentUrl));
```

aduh95 commented on lib/internal/modules/esm/loader.js:
```diff
@@ -513,19 +519,44 @@ function getHooksProxy() {
+function register(specifier, parentURL = undefined, options) {
   const moduleLoader = require('internal/process/esm_loader').esmLoader;
-  moduleLoader.register(`${specifier}`, parentURL);
+  if (parentURL !== null && typeof parentURL === 'object') {
```
nit: we could have a shortcut for `undefined` since that’s the default value. 
```suggestion
  if (parentURL != null && typeof parentURL === 'object') {
```

targos commented on doc/api/module.md:
````diff
@@ -175,6 +175,28 @@ globalPreload: http-to-https
 globalPreload: unpkg
 ```
 
+This function can also be used to pass data to the loader's [`initialize`][]
+hook include transferrable objects like ports.
````
This sentence seems weird to me (non-native English speaker)

izaakschroeder replied:
Oops `include` should be `including` 😅

GeoffreyBooth commented:
@aduh95 I think this PR can rename `parentUrl` to `parentURL`. Arguably the title of the PR should be something more like “esm: add `initialize` hook, improve `register`” as there’s significant refactoring to `register` happening here, and renaming one of `register`‘s parameters could easily be part of that.

mcollina reviewed (approved):
lgtm

aduh95 commented:
@izaakschroeder can you please rebase to fix the git conflict?

izaakschroeder commented:
Done @aduh95 🎉

GeoffreyBooth commented on doc/api/esm.md:
```diff
@@ -739,6 +742,62 @@ different [realm](https://tc39.es/ecma262/#realm). The hooks thread may be
 terminated by the main thread at any time, so do not depend on asynchronous
 operations (like `console.log`) to complete.
 
+#### `initialize()`
```
I think he means that you should update the docs to clarify those points. And if the function is allowed to be async, make the example function an async function (it doesn’t need to actually await anything, just put the `async` keyword in the signature).

izaakschroeder replied:
Done!

GeoffreyBooth commented on doc/api/esm.md:
```diff
@@ -949,8 +1008,8 @@ changes:
-> The loaders API is being redesigned. This hook may disappear or its
-> signature may change. Do not rely on the API described below.
+> This hook will be removed in a future version. Use [`initialize`][] instead.
+> When a loader has an `initilize` export, `globalPreload` will be ignored.
```
```suggestion
> When a loader has an `initialize` export, `globalPreload` will be ignored.
```

izaakschroeder commented:
Anything else left @GeoffreyBooth @aduh95? 😄

GeoffreyBooth commented:
> Anything else left @GeoffreyBooth @aduh95? 😄 

Do you mind changing the first commit message to something like:

``` esm: add `initialize` hook, integrate with `register`
```

So that the commit message on `main` mentions that this changes `register`.

izaakschroeder commented:
> Do you mind changing the first commit message to something like:

Can do 🎉 Should I then rebase and squash everything as well?

GeoffreyBooth commented:
> Can do 🎉 Should I then rebase and squash everything as well?

No, the landing script will do that.

izaakschroeder commented:
Reworded 😄

giltayar commented:
Will this land in the next v20?

GeoffreyBooth commented:
@nodejs/releasers, please ensure that https://github.com/nodejs/node/pull/46826, https://github.com/nodejs/node/pull/48559 and https://github.com/nodejs/node/pull/48842 all go out together in the same release.

ruyadorno commented:
Based on the note from @GeoffreyBooth on https://github.com/nodejs/node/pull/46826#issuecomment-1663314418 I'm going ahead and adding a backport-requested label to this PR too. Ideally you can organize it so that it can be a single backport PR containing all the required commits 😊
